File: aoc-2025/cpp/src/day11.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

const std::string DEFAULT_INPUT = "../input/day11.txt";
const std::string DEFAULT_OUTPUT = "output/day11.txt";

using Graph = std::unordered_map<std::string, std::vector<std::string>>;
using Memo = std::unordered_map<std::string, long long>;

Graph parse_input(const std::string &filename) {
  std::ifstream file(filename);
  Graph adj;
  if (!file.is_open())
    return adj;

  std::string line;
  while (std::getline(file, line)) {
    if (line.empty())
      continue;

    size_t colon_pos = line.find(':');
    if (colon_pos == std::string::npos)
      continue;

    std::string u = line.substr(0, colon_pos);
    std::string rem = line.substr(colon_pos + 1);

    std::stringstream ss(rem);
    std::string v;
    while (ss >> v)
      adj[u].push_back(v);
  }

  return adj;
}

long long count_paths(const std::string &curr, const std::string &target,
                      Graph &adj, Memo &memo) {
  if (curr == target)
    return 1;

  if (memo.count(curr))
    return memo[curr];

  long long ans = 0;

  if (adj.count(curr))
    for (auto &nbr : adj[curr])
      ans += count_paths(nbr, target, adj, memo);

  return memo[curr] = ans;
}

long long get_paths(const std::string &start, const std::string &end,
                    Graph &adj) {
  Memo memo;
  return count_paths(start, end, adj, memo);
}

long long part_one(const std::string &filename) {
  Graph adj = parse_input(filename);
  if (adj.empty())
    return -1;
  return get_paths("you", "out", adj);
}

long long part_two(const std::string &filename) {
  Graph adj = parse_input(filename);
  if (adj.empty())
    return -1;

  long long p1_leg1 = get_paths("svr", "dac", adj);
  long long p1_leg2 = get_paths("dac", "fft", adj);
  long long p1_leg3 = get_paths("fft", "out", adj);
  long long path1 = p1_leg1 * p1_leg2 * p1_leg3;

  long long p2_leg1 = get_paths("svr", "fft", adj);
  long long p2_leg2 = get_paths("fft", "dac", adj);
  long long p2_leg3 = get_paths("dac", "out", adj);
  long long path2 = p2_leg1 * p2_leg2 * p2_leg3;

  return path1 + path2;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  auto p1 = part_one(input_file);
  auto p2 = part_two(input_file);

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << p1 << "\n" << p2 << "\n";

  return 0;
}

File: aoc-2025/rust/src/bin/day11.rs
//! https://adventofcode.com/2025/day/11
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "";
const DEFAULT_OUTPUT: &str = "";

type Graph = HashMap<String, Vec<String>>;
type Memo = HashMap<String, u64>;

fn parse_input(content: &str) -> Graph {
    let mut graph = HashMap::new();
    for line in content.lines() {
        if line.trim().is_empty() {
            continue;
        }
        if let Some((src, dests)) = line.split_once(':') {
            let src = src.trim().to_string();
            let neighbors: Vec<String> = dests.split_whitespace().map(|s| s.to_string()).collect();
            graph.insert(src, neighbors);
        }
    }
    graph
}

fn count_paths(curr: &str, target: &str, graph: &Graph, memo: &mut Memo) -> u64 {
    if curr == target {
        return 1;
    }
    if let Some(&count) = memo.get(curr) {
        return count;
    }

    let mut total = 0;
    if let Some(neighbors) = graph.get(curr) {
        for next in neighbors {
            total += count_paths(next, target, graph, memo);
        }
    }

    memo.insert(curr.to_string(), total);
    total
}

fn get_paths(start: &str, end: &str, graph: &Graph) -> u64 {
    let mut memo = HashMap::new();
    count_paths(start, end, graph, &mut memo)
}

fn part_one(content: &str) -> u64 {
    let graph = parse_input(content);
    get_paths("you", "out", &graph)
}

fn part_two(content: &str) -> u64 {
    let graph = parse_input(content);

    let path_dac_fft = get_paths("svr", "dac", &graph)
        * get_paths("dac", "fft", &graph)
        * get_paths("fft", "out", &graph);

    let path_fft_dac = get_paths("svr", "fft", &graph)
        * get_paths("fft", "dac", &graph)
        * get_paths("dac", "out", &graph);

    path_dac_fft + path_fft_dac
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = part_one(&content);
    let p2 = part_two(&content);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}
